File: src/main/java/io/stouder/adonis/cli/json/routes/RouteHandler.java
package io.stouder.adonis.cli.json.routes;

import com.google.gson.*;
import lombok.Getter;

import java.lang.reflect.Type;

@Getter
public class RouteHandler {
    private String type;

    public static class Deserializer implements JsonDeserializer<RouteHandler> {
        @Override
        public RouteHandler deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
            JsonObject jsonObject = jsonElement.getAsJsonObject();
            JsonElement jsonType = jsonObject.get("type");
            String typeString = jsonType.getAsString();
            Class<? extends RouteHandler> handlerClass = switch (typeString) {
                case "controller" -> ControllerRouteHandler.class;
                case "closure" -> ClosureRouteHandler.class;
                default -> throw new JsonParseException("Unknown type: " + typeString);
            };

            return jsonDeserializationContext.deserialize(jsonElement, handlerClass);
        }
    }
}

File: src/main/java/io/stouder/adonis/cli/json/routes/RouteHandler.kt
package io.stouder.adonis.cli.json.routes

import com.google.gson.*
import java.lang.reflect.Type

open class RouteHandler(
    val type: String
) {
    class Deserializer : JsonDeserializer<RouteHandler> {
        override fun deserialize(jsonElement: JsonElement, type: Type, jsonDeserializationContext: JsonDeserializationContext): RouteHandler {
            val jsonObject = jsonElement.asJsonObject
            val jsonType = jsonObject.get("type")
            val handlerClass: Class<out RouteHandler> = when (val typeString = jsonType.asString) {
                "controller" -> ControllerRouteHandler::class.java
                "closure" -> ClosureRouteHandler::class.java
                else -> throw JsonParseException("Unknown type: $typeString")
            }

            return jsonDeserializationContext.deserialize(jsonElement, handlerClass)
        }
    }
}
